[Persistence] Adds Save and Load functionality to TreeStore

dylanlott opened the pull request:
## Description

Adds save and load functionality to the TreeStore.

This PR makes the `TreeStore` an `Interruptable` module. This is necessary because the Persistence module needs to be able to tell the `TreeStore` to relinquish its connection to the `nodeStores` which are backed by BadgerDB.

## Issue

Related to #568

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes


- Adds save and load functionality for loading a `worldState` from file directories.

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

Olshansk reviewed (changes requested):
A few comments throughout but nothing "glaring" or blocking.

Thanks for getting this moving Dylan!

dylanlott commented on persistence/trees/trees_test.go:
```diff
@@ -50,8 +50,6 @@ func TestTreeStore_Update(t *testing.T) {
 		pmod := newTestPersistenceModule(t, dbUrl)
 		context := newTestPostgresContext(t, 0, pmod)
 
-		require.NoError(t, context.SetSavePoint())
```
They're an unnecessary holdover; ComputeStateHash does the update and commit of the test.

dylanlott commented on persistence/kvstore/kvstore.go:
```diff
@@ -22,6 +25,8 @@ type KVStore interface {
 	GetAll(prefixKey []byte, descending bool) (keys, values [][]byte, err error)
 	Exists(key []byte) (bool, error)
 	ClearAll() error
+
```
I added `Accessors` let me know if you think of anything better, I couldn't think of anything better either. Lifecycle didn't feel accurate.

Olshansk commented:
@dylanlott Can you "re-request" a review when its ready and I'll reply to all the comments/questions then?

dylanlott commented on persistence/trees/atomic_test.go:
```diff
@@ -89,4 +91,132 @@ func TestTreeStore_AtomicUpdatesWithSuccessfulRollback(t *testing.T) {
+	return ts
+}
+
+func isEmpty(dir string) (bool, error) {
```
Addressed with a tech debt comment to #796 👍

dylanlott commented:
> @dylanlott Can you "re-request" a review when its ready and I'll reply to all the comments/questions then?

You beat me to it. I just finished up addressing your feedback, it's ready for re-review 👍

dylanlott commented on persistence/trees/module_test.go:
```diff
@@ -1,55 +1,53 @@
-	serviceURLFormat = "node%d.consensus:42069"
-)
+// DISCUSS: This is duplicated from inside trees package. Is it worth exporting or is it better as duplicate code?
+var stateTreeNames = []string{
```
I'm going to file this change in a different PR, but I agree with it. However, it is more complicated than belongs in this PR, if you're cool with that. Would you like me to leave a comment or should I just file the PR?
